### PR TITLE
Minor updates to ch2 based on feedback

### PR DESCRIPTION
**Pull request by jni**
Note specifically the use of `np.apply_along_axis`, which I hadn't heard of until I started looking into how to use `stats.rankdata` instead of `stats.mstats.rankdata`! =) This version is 3x faster.

### CODE REVIEW

File: markdown/ch2.markdown [hunk 1]
```diff
@@ -77,52 +95,59 @@ import pandas as pd
 
 # Import TCGA melanoma data
 filename = 'data/counts.txt'
-with open(filename, 'rt') as f:
-    data_table = pd.read_csv(f, index_col=0) # Parse file with pandas
+data_table = pd.read_csv(filename, index_col=0) # Parse file with pandas
```
Comment (stefanv) [inline markdown/ch2.markdown]: PEP8  (two spaces before #)

File: markdown/ch2.markdown [hunk 2]
```diff
@@ -77,52 +95,59 @@ import pandas as pd
...
+First, as usual, we set our plotting style:
+
 ```python
 # Make plots appear inline, set custom plotting style
 %matplotlib inline
 import matplotlib.pyplot as plt
 plt.style.use('style/elegant.mplstyle')
 ```
 
+Next, we write a plotting function that makes use of scipy's `gaussian_kde` function to plot smooth distributions:
```
Comment (stefanv) [inline markdown/ch2.markdown]: SciPy?

File: markdown/ch2.markdown [hunk 3]
```diff
@@ -538,14 +564,20 @@ clusters = fcluster(yc, threshold_distance, 'distance')
...
-This is indeed only the latest study to show such a result, with others identifying subtypes of leukemia (blood cancer), gut cancer, and more.
-Although the above clustering technique is quite fragile, there are other ways to explore this dataset and similar ones that are more robust.
+The clustering of gene expression profiles appears to have identified a
+higher-risk subtype of melanoma, which constitutes the majority of patients.
+The TCGA study backs this claim up with a more robust clustering and
+statistical testing.  This is indeed only the latest study to show such a
+result, with others identifying subtypes of leukemia (blood cancer), gut
+cancer, and more.  Although the above clustering technique is quite fragile,
+there are other ways to explore this dataset and similar ones that are more
+robust.
```
Comment (stefanv) [inline markdown/ch2.markdown]: Reference?

File: markdown/ch2.markdown (hunk 2 above)
Comment (jni) [inline markdown/ch2.markdown]: Oops! Fixed.

Or we can just rename the whole thing to "elegant scipy". =P

**Comment by jni** [comment]
@stefanv did you know about `np.apply_along_axis`??? This function will change my life! =D

**Comment by stefanv** [comment]
Heh, you'll remember my little NumPy exercise in 2008:

```
7f8aae03 numpy/lib/shape_base.py  (Albert Jornet Puig  2014-06-06 10:40:32 +0200  20) def apply_along_axis(func1d, axis, arr, *args, **kwargs):
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  21)     """
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  22)     Apply a function to 1-D slices along the given axis.
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  23) 
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  24)     Execute `func1d(a, *args)` where `func1d` operates on 1-D arrays and `a`
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  25)     is a 1-D slice of `arr` along `axis`.
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  26) 
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  27)     Parameters
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  28)     ----------
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  29)     func1d : function
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  30)         This function should accept 1-D arrays. It is applied to 1-D
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  31)         slices of `arr` along the specified axis.
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  32)     axis : integer
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  33)         Axis along which `arr` is sliced.
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  34)     arr : ndarray
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  35)         Input array.
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  36)     args : any
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  37)         Additional arguments to `func1d`.
61a9a949 numpy/lib/shape_base.py  (Michael Seifert     2016-02-18 14:21:14 +0100  38)     kwargs : any
ddc95d37 numpy/lib/shape_base.py  (Albert              2014-03-06 16:17:52 +0100  39)         Additional named arguments to `func1d`.
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  40) 
7f8aae03 numpy/lib/shape_base.py  (Albert Jornet Puig  2014-06-06 10:40:32 +0200  41)         .. versionadded:: 1.9.0
7f8aae03 numpy/lib/shape_base.py  (Albert Jornet Puig  2014-06-06 10:40:32 +0200  42) 
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  43) 
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  44)     Returns
6647bf7e numpy/lib/shape_base.py  (Stefan van der Walt 2008-08-05 09:20:07 +0000  45)     -------
f4dd54aa numpy/lib/shape_base.py  (Ralf Gommers        2012-03-03 21:20:13 +0100  46)     apply_along_axis : ndarray
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  47)         The output array. The shape of `outarr` is identical to the shape of
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  48)         `arr`, except along the `axis` dimension, where the length of `outarr`
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  49)         is equal to the size of the return value of `func1d`.  If `func1d`
7b751f66 numpy/lib/shape_base.py  (Pauli Virtanen      2009-03-24 22:25:21 +0000  50)         returns a scalar `outarr` will have one fewer dimensions than `arr`.
```

**Comment by jni** [comment]
Mind. Blown. =D

Why was this not in your Advanced NumPy slides/exercises? ;)

**Comment by jni** [comment]
@stefanv @hdashnow Adding an exercise using `np.apply_along_axis` might be awesome. (But let's not hold up this PR for that.)

**Comment by hdashnow** [comment]
Nice!

**Pull request merged**